src/box/xrow_io.cc

```diff
@@ -102,3 +105,54 @@ coio_write_xrow(struct iostream *io, const struct xrow_header *row)
 		diag_raise();
 }
 
+uint64_t xrow_stream_flush_size = 16384;
+TWEAK_UINT(xrow_stream_flush_size);
+
+void
+xrow_stream_write(struct xrow_stream *stream, const struct xrow_header *row)
+{
+	/* Fixheader - length of packet. */
+	size_t fixheader_len = mp_sizeof_uint(UINT32_MAX);
+	size_t len = 0;
+	len += xrow_header_sizeof(row, row->sync);
+	for (int i = 0; i < row->bodycnt; i++) {
+		len += row->body[i].iov_len;
+	}
+	char *data = (char *)xlsregion_alloc(&stream->lsregion,
+					     len + fixheader_len,
+					     ++stream->lsr_id);
+	*(data) = 0xce; /* MP_UINT32 */
+	store_u32(data + 1, mp_bswap_u32(len));
+	data = data + fixheader_len;
+	data += xrow_header_encode(row, row->sync, data);
+	for (int i = 0; i < row->bodycnt; i++) {
+		size_t l = row->body[i].iov_len;
+		memcpy(data, row->body[i].iov_base, l);
+		data += l;
+	}
+}
+
+int
+xrow_stream_flush(struct xrow_stream *stream, struct iostream *io)
+{
+	ssize_t to_flush = lsregion_used(&stream->lsregion);
+	/*
+	 * Might flush more than requested if data is added to the buffer
+	 * during the coio_writev yield.
+	 */
+	while (to_flush > 0) {
+		struct iovec iov[IOV_MAX];
+		int iovcnt = lengthof(iov);
+		int64_t gc_id = lsregion_to_iovec(&stream->lsregion, iov,
+						  &iovcnt, &stream->flush_pos);
+		ssize_t written = coio_writev(io, iov, iovcnt, 0);
+		if (written < 0)
+			return -1;
+		to_flush -= written;
+		lsregion_gc(&stream->lsregion, gc_id);
+	}
+	struct errinj *inj = errinj(ERRINJ_RELAY_TIMEOUT, ERRINJ_DOUBLE);
+	if (inj != NULL && inj->dparam > 0)
+		fiber_sleep(inj->dparam);
+	return 0;
```

Comment:
Can this errinj somehow stay in relay?

Reply:
Ok:
<details><summary>diff</summary>

```diff
diff --git a/src/box/relay.cc b/src/box/relay.cc
index fa9d8e500a..405a9c437a 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -442,6 +442,34 @@ relay_cord_init(struct relay *relay)
 	xrow_stream_create(&relay->xrow_stream);
 }
 
+/** Flush any relay stream contents to the remote peer immediately. */
+static inline int
+relay_flush(struct relay *relay)
+{
+	if (xrow_stream_flush(&relay->xrow_stream, relay->io) < 0)
+		return -1;
+#ifndef NDEBUG
+	struct errinj *inj = errinj(ERRINJ_RELAY_TIMEOUT, ERRINJ_DOUBLE);
+	if (inj != NULL && inj->dparam > 0)
+		fiber_sleep(inj->dparam);
+#endif
+	return 0;
+}
+
+/** Check if relay stream has enough data to flush and flush it. */
+static inline int
+relay_check_flush(struct relay *relay)
+{
+	if (xrow_stream_check_flush(&relay->xrow_stream, relay->io) < 0)
+		return -1;
+#ifndef NDEBUG
+	struct errinj *inj = errinj(ERRINJ_RELAY_TIMEOUT, ERRINJ_DOUBLE);
+	if (inj != NULL && inj->dparam > 0)
+		fiber_sleep(inj->dparam);
+#endif
+	return 0;
+}
+
 void
 relay_initial_join(struct iostream *io, uint64_t sync, struct vclock *vclock,
 		   uint32_t replica_version_id)
@@ -518,7 +546,7 @@ relay_initial_join(struct iostream *io, uint64_t sync, struct vclock *vclock,
 
 	/* Send read view to the replica. */
 	engine_join_xc(&ctx, &relay->stream);
-	if (xrow_stream_flush(&relay->xrow_stream, io) < 0)
+	if (relay_flush(relay) < 0)
 		diag_raise();
 }
 
@@ -534,7 +562,7 @@ relay_final_join_f(va_list ap)
 	assert(relay->stream.write != NULL);
 	recover_remaining_wals(relay->r, &relay->stream,
 			       &relay->stop_vclock, true);
-	if (xrow_stream_flush(&relay->xrow_stream, relay->io) < 0)
+	if (relay_flush(relay) < 0)
 		diag_raise();
 	assert(vclock_compare(&relay->r->vclock, &relay->stop_vclock) == 0);
 	return 0;
@@ -826,7 +854,7 @@ relay_send_heartbeat(struct relay *relay)
 	row.replica_id = instance_id;
 	relay->last_heartbeat_time = ev_monotonic_now(loop());
 	relay_send(relay, &row);
-	if (xrow_stream_flush(&relay->xrow_stream, relay->io) < 0) {
+	if (relay_flush(relay) < 0) {
 		relay_set_error(relay, diag_last_error(diag_get()));
 		fiber_cancel(fiber());
 	}
@@ -1066,7 +1094,7 @@ relay_subscribe_f(va_list ap)
 					 relay->last_row_time + timeout);
 		cbus_process(&relay->wal_endpoint);
 		relay_subscribe_update(relay);
-		if (xrow_stream_flush(&relay->xrow_stream, relay->io) < 0) {
+		if (relay_flush(relay) < 0) {
 			relay_set_error(relay, diag_last_error(diag_get()));
 			fiber_cancel(fiber());
 		}
@@ -1162,7 +1190,7 @@ relay_send_initial_join_row(struct xstream *stream, struct xrow_header *row)
 	 */
 	if (row->group_id != GROUP_LOCAL)
 		relay_send(relay, row);
-	if (xrow_stream_check_flush(&relay->xrow_stream, relay->io) < 0)
+	if (relay_check_flush(relay) < 0)
 		diag_raise();
 }
 
@@ -1179,7 +1207,7 @@ relay_raft_msg_push(struct cmsg *base)
 	RegionGuard region_guard(&fiber()->gc);
 	xrow_encode_raft(&row, &fiber()->gc, &msg->req);
 	relay_send(relay, &row);
-	if (xrow_stream_flush(&relay->xrow_stream, relay->io) < 0) {
+	if (relay_flush(relay) < 0) {
 		relay_set_error(relay, diag_last_error(diag_get()));
 		fiber_cancel(fiber());
 	}
@@ -1338,7 +1366,7 @@ relay_send_tx(struct relay *relay)
 		}
 		relay_send(relay, packet);
 	}
-	if (xrow_stream_check_flush(&relay->xrow_stream, relay->io) < 0)
+	if (relay_check_flush(relay) < 0)
 		diag_raise();
 
 	rlist_create(&relay->current_tx);
diff --git a/src/box/xrow_io.cc b/src/box/xrow_io.cc
index 0dae40b60a..696032e43b 100644
--- a/src/box/xrow_io.cc
+++ b/src/box/xrow_io.cc
@@ -33,7 +33,6 @@
 #include "xrow.h"
 #include "coio.h"
 #include "coio_buf.h"
-#include "errinj.h"
 #include "error.h"
 #include "msgpuck/msgpuck.h"
 #include "tweaks.h"
@@ -151,8 +150,5 @@ xrow_stream_flush(struct xrow_stream *stream, struct iostream *io)
 		to_flush -= written;
 		lsregion_gc(&stream->lsregion, gc_id);
 	}
-	struct errinj *inj = errinj(ERRINJ_RELAY_TIMEOUT, ERRINJ_DOUBLE);
-	if (inj != NULL && inj->dparam > 0)
-		fiber_sleep(inj->dparam);
 	return 0;
 }

```

<details>